### PR TITLE
ecr credential plugin

### PR DESCRIPTION
**Pull request by ibrokethecloud**
Added logic for handling ecr token generation within RKE.

This seems to be a common requirements for some customers spinning up RKE clusters outside of AWS but using AWS ECR, where the AWS cloud provider can not be used.

Coupled with the [amazon-ecr-credential-helper](https://github.com/awslabs/amazon-ecr-credential-helper) on the cluster nodes, the cluster should be able to pull images on nodes not running in AWS but using ECR.

The PR allows the user to authenticate with AWS using their credentials stored on the nodes or provided statically in the manifest.

Following is a sample of the same. 

```
services:
  kubelet:
    extra_env:
      - "AWS_ACCESS_KEY_ID=your_access_key"
      - "AWS_SECRET_ACCESS_KEY=your_secret_key"         
private_registries:
     - url: account-id.dkr.ecr.region.amazonaws.com
       is_default: true
       credentialPlugin:
        type: ecr
        # Static Credentials
        #aws_access_key_id: your_access_key
        #aws_secret_access_key: your_secret_key
        #aws_session_token: your_session_token
     - url: myprivatereg
       user: username
       password: password
```

For environments outside AWS, the kubelet will need access to AWS credentials to be able to leverage the ecr-plugin available on the cluster nodes. 

The change generates a the config.json for the kubelet to be of the form:
```
{"auths":{"myprivatereg":{"auth":"dXNlcm5hbWU6cGFzc3dvcmQ="}},"credHelpers":{"account-id.dkr.ecr.region.amazonaws.com":"ecr-login"}}
```

Related to known issues: 
https://github.com/rancher/rancher/issues/28693
https://github.com/rancher/rke/issues/1795

### CODE REVIEW

**Comment by tabnul** [comment]
any update on this one? waiting for this exact functionality :)

**Comment by superseb** [comment]
@ibrokethecloud This is needed for the initial pull and the helper will refresh the token? Was it tested that the token was refreshed after the initial 12 hours?

Can you also rebase it?

**Comment by ibrokethecloud** [comment]
@superseb i will run a longer test, the kubelet should refresh the token, but i will confirm and will also rebase the changes.

**Comment by joehoeller** [comment]
@superseb any word on this?

**Comment by superseb** [comment]
The word is that @ibrokethecloud is doing final testing and we are awaiting feedback :)

**Comment by joehoeller** [comment]
Cool, thank you 🙏🏻 😎

**Comment by joehoeller** [comment]
Once this is in, can you please leave instructions on how to add this to an
RKE cluster?

**Comment by ibrokethecloud** [comment]
@superseb I ran a test, and the kubelet can still pull all the images nearly 24 hours later.

Here is the sample cronjob.

```
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: hello
spec:
  schedule: "0 * * * *"
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - name: hello
            image: accountID.dkr.ecr.ap-southeast-2.amazonaws.com/ubuntu:latest
            imagePullPolicy: Always
            command:
            - /bin/sh
            - -c
            - date; echo Hello from the Kubernetes cluster
          restartPolicy: OnFailure
```

```
▶ k get cronjob
NAME    SCHEDULE    SUSPEND   ACTIVE   LAST SCHEDULE   AGE
hello   0 * * * *   False     0        36m             20h
(⎈ |local:default)
work/repos/ecrtest
▶ k get pods
NAME                     READY   STATUS      RESTARTS   AGE
hello-1624053600-nt2dz   0/1     Completed   0          156m
hello-1624057200-g4l9p   0/1     Completed   0          96m
hello-1624060800-qtmbx   0/1     Completed   0          36m
hello-kbph4              0/1     Completed   0          20h
```

To get started just need to ensure that there is the ecr-credential-helper is installed on all nodes on the cluster.  Instructions for the same are available [here](https://github.com/awslabs/amazon-ecr-credential-helper).

I used a IAM policy with scoped access to my ECR repo. Sample one is available here..

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": [
                "ecr:DescribeImageScanFindings",
                "ecr:GetLifecyclePolicyPreview",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
                "ecr:DescribeImages",
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetRepositoryPolicy",
                "ecr:GetLifecyclePolicy"
            ],
            "Resource": "arn:aws:ecr:ap-southeast-2:ACCOUNT_ID:repository/ubuntu"
        },
        {
            "Sid": "VisualEditor1",
            "Effect": "Allow",
            "Action": [
                "ecr:GetRegistryPolicy",
                "ecr:DescribeRegistry",
                "ecr:GetAuthorizationToken"
            ],
            "Resource": "*"
        }
    ]
}
```

This needs to be associated with an IAM user, and the API access / secret tokens can then  be used in the cluster.yml

```
nodes:
  - address: node1
    user: root
    role:
      - controlplane
      - etcd
      - worker
services:
  kubelet:
    extra_env:
      - "AWS_ACCESS_KEY_ID=TOKEN"
      - "AWS_SECRET_ACCESS_KEY=PASSWORD"
private_registries:
     - url: ACCOUNT_ID.dkr.ecr.ap-southeast-2.amazonaws.com
       is_default: false
       credentialPlugin:
        type: ecr
```

I hope this helps, in case anyone else wants to test it out.

I will rebase the changes.

File: docker/docker.go [hunk 1]
```diff
@@ -738,12 +759,22 @@ func isContainerEnvChanged(containerEnv, imageConfigEnv, dockerfileEnv []string)
...
+	credHelper := make(map[string]string)
 	for url, pr := range prsMap {
-		auth := base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%s:%s", pr.User, pr.Password)))
-		auths[url] = authConfig{Auth: auth}
+		if len(pr.CredentialPlugin) != 0 {
+			if credPluginType, ok := pr.CredentialPlugin["type"]; ok {
+				if credPluginType == "ecr" {
+					credHelper[pr.URL] = "ecr-login"
+				} else {
+					logrus.Infof("Skipping plugin in kubelet config.json. Plugin %s unsupported", credPluginType)
```
Comment (superseb) [inline docker/docker.go:769]: Can we move this to `cluster/validation.go` and error out as its a non supported config.

Comment (ibrokethecloud) [reply docker/docker.go]: @superseb i have re-organized the code

**Review by superseb — APPROVED** [review]

**Pull request merged**